docs/upgrading.md

```diff
@@ -239,3 +262,26 @@ Your code should be adapted to manipulate tokens like this:
 -echo (string) $token;
 +echo $token->toString();
 ```
+
+### An incompatible change related to the `aud` claim handling
+
+In 3.4 the handling of audience claims has changed.
+
+The new behaviour in 4.0 is to allow easier access to the `aud` claims by always representing it as an array with 
+only one entry if the original token has a string, so `"string"` will be converted to `["string"]`. In 4.0 you can 
+influence this by adding or removing the `\Lcobucci\JWT\Encoding\UnifyAudience` claims formatter class, or your own 
+implementation.
+
+In 3.4 however there is no such thing available, and the implementation hardcoded casts the argument passed to
+`\Lcobucci\JWT\Builder::permittedFor()` as a string (forcing an array to become the string "Array" and triggering the
+"array to string conversion" PHP notice), wrapping it inside an array, with no way to add more than one string to the 
+token. Reading from the token via `\Lcobucci\JWT\Token::getClaim()` or`\Lcobucci\JWT\Token::getClaims()` will revert 
+this array wrapping and only ever return the first entry in the "aud" claims array, if it is one, or return the string 
+otherwise.
+
+If you pass an array to the builder or if you read an array from the token, a deprecation message will be
+triggered to inform you of this behaviour change. There is no workaround for version 3.4 if you have to
+access or create "aud" claims that are arrays with multiple entries.
+
+In 4.0, the `permittedFor()` method accepts a variadic parameter, so you can either call 
+`permittedFor("audience1", "audience2")` or `permittedFor(...["audience1", "audience2"])` 
```

Comment:
This assumes that the library had official support for multiple audiences on v3, which is not the case.

I'm happy to improve the docs to mention that things accidentally used to work because of lack of type declaration and loose validation on custom claim configuration, though.

Reply:
I think I found the hidden reason, and illustrated it in the new version of this part.